packages/std/src/imports.rs

```diff
@@ -25,7 +25,7 @@ extern "C" {
 
     // scan creates an iterator, which can be read by consecutive next() calls
     #[cfg(feature = "iterator")]
-    fn db_scan(start: *const c_void, end: *const c_void, order: i32) -> i32;
+    fn db_scan(start: *const c_void, end: *const c_void, order: i32) -> u32;
```

Comment:
But ExternalIterator is a struct, not a number. This struct does not exist on the VM side and the VM does not create it. What we could do is to make it a newtype of u32.

In general I want to avoid padding around pointers of complex types. This always requires that both side know the type and use the same memory layout for it. I'm quite happy we only need numbers and Region pointers.

One thing we could do for `db_read`, `db_next`, `canonicalize_address`, `humanize_address` and `query_chain` is to change the return type to `*mut Region`. The problem I'm seeing here is that in the VM, the type is not `*mut Region` so we'd feed we provide a u32 result to an import that expacts a `*mut Region`, which then is comiled down to u32. I'm not convinced that reducing expliciteness helps here.

Reply:
It's really a question of explicitness, i agree, but you have already made this same choice for the import function arguments, which take pointer, which are interpreted as `u32` in the VM. When pairs of types have the same size, layout, alignment, and set of legal values, then you can safely transmute between them. That means that you can write to a `u32` on one side, and interpret it as another type on the other. The declaration that i had in mind is:
```rust
#[repr(C)]
struct ExternalIterator {
    iterator_id: u32,
}
```
which would have the same memory representation as the `u32`, but gets the same benefits as a regular type. (The only difference between this and the original definition is the `repr` line

Reply:
> but you have already made this same choice for the import function arguments, which take pointers, which are interpreted as `u32` in the VM

Yeah – we were young and did not know what we were doing.

I prefer to use simple types on imports and exports that directly translate to Wasm types. This makes it much easier to add support for other smart contract languages than Rust, since you don't need to think about memory layout for advances types other than `Region`.

Reply:
> Yeah – we were young and did not know what we were doing.

:laughing: 

Yes plenty of learning working over this FFI stuff again and again.
I'm actually kind of amazed this worked without any panics back in 0.5 when I see how many open questions there are.

Reply:
https://github.com/CosmWasm/cosmwasm/issues/424